pixeltable/exec/exec_context.py

```diff
@@ -1,42 +1,210 @@
+import builtins
+import logging
 import random
+import time
 
 import sqlalchemy as sql
+from rich.progress import Column, Progress, ProgressColumn, Task, TaskID, Text, TextColumn
 
 from pixeltable import exprs
+from pixeltable.env import Env
+
+_logger = logging.getLogger('pixeltable')
 
 
 class ExecContext:
     """Class for execution runtime constants"""
 
     row_builder: exprs.RowBuilder
+    show_progress: bool
+    progress: Progress | None
+    progress_start: float  # time.monotonic() of progress.start()
+    progress_reporters: dict[str, 'ExecContext.ProgressReporter']
+    elapsed_time_task_id: TaskID | None
+    batch_size: int  # 0: no batching
     profile: exprs.ExecProfile
-    show_pbar: bool
-    batch_size: int
-    num_rows: int | None
-    conn: sql.engine.Connection | None
+    conn: sql.engine.Connection | None  # if present, use this to execute SQL queries
     pk_clause: list[sql.ClauseElement] | None
-    num_computed_exprs: int
     ignore_errors: bool
     random_seed: int  # general-purpose source of randomness with execution scope
 
+    class ProgressReporter:
```

Comment:
Does it have to be a nested class? Seems like it would be more readable as a sister class in the same module. Unless nesting it gives it a special meaning.